File: 8주차/구민지/행렬 테두리 회전하기.py
import copy

def solution(rows, columns, queries):
    answer = []
    graph=[[r*columns+c for c in range(1,columns+1)] for r in range(rows)]
    new_graph=[[0]*columns for _ in range(rows)]

    INF=int(1e9)
    for query in queries:
        r1,c1,r2,c2=query
        graph2=[[0]*columns for _ in range(rows)]
        min_num=INF
        for r in range(rows):
            for c in range(columns):
                if r1-1<=r<=r2-1 and c1-1<=c<=c2-1:
                    if r1<=r<r2-1 and c1<=c<c2-1:
                        graph2[r][c]=graph[r][c]
                        continue
                    if graph[r][c]<min_num:
                        min_num=graph[r][c]
                    if c1-1<=c<c2-1 and r==r1-1:
                        graph2[r][c+1]=graph[r][c]
                    elif c==c2-1 and r1-1<=r<r2-1:
                        graph2[r+1][c]=graph[r][c]
                    elif r==r2-1 and c1-1<c<=c2-1:
                        graph2[r][c-1]=graph[r][c]
                    else:
                        graph2[r-1][c]=graph[r][c]
                    continue
                graph2[r][c]=graph[r][c]
        graph=graph2
        answer.append(min_num)

    return answer

File: 8주차/김용상/행렬테두리회전하기.java
class Solution {
    public int[] solution(int rows, int columns, int[][] queries) {
        int[] answer = new int[queries.length];
        // [1][1] 값을 tmp하나에 옮기기
        // 4번의 작업을 통해 옮겨야함.
        int[][]map = new int[rows+1][columns+1];
        int cnt = 1;
        for (int i=1;i<=rows;i++){
            for (int j=1;j<=columns;j++){
                map[i][j] = cnt++; 
            }
        }
        int i=0;
        for (int[]query :queries){
            int tmp = 0;
            int start_x = query[0];
            int start_y = query[1];
            int end_x = query[2];
            int end_y = query[3];
            //상
            tmp = map[start_x][end_y];
            int mn = tmp;
            for (int k=end_y;k>start_y;k--){
                map[start_x][k] = map[start_x][k-1];
                if (map[start_x][k]<mn)
                    mn = map[start_x][k];
            }
            //좌
            for (int k=start_x;k<end_x;k++){
                map[k][start_y] = map[k+1][start_y];
                if(map[k][start_y]<mn)
                    mn = map[k][start_y];
            }
            //하
            for (int k=start_y;k<end_y;k++){
                map[end_x][k]= map[end_x][k+1];
                if (map[end_x][k]<mn)
                    mn = map[end_x][k];
            }
            //우
            for (int k=end_x;k>start_x+1;k--){
                map[k][end_y] = map[k-1][end_y];
                if (map[k][end_y]<mn)
                    mn = map[k][end_y];
            }
            //tmp는 직접
            map[start_x+1][end_y] = tmp; 
            answer[i++] = mn;

        }
        
        return answer;
    }
}
